src/NHibernate/Impl/AbstractQueryImpl.cs

```diff
@@ -674,7 +676,34 @@ public IQuery SetProperties(IDictionary map)
 			return this;
 		}
 
+		public IQuery SetParameters(IDictionary<string, object> map)
+		{
+			foreach (var namedParam in NamedParameters)
+			{
+				var obj = map[namedParam];
+				switch (obj)
+				{
+					case null:
+						break;
+					case IEnumerable enumerable when !(enumerable is string):
+						SetParameterList(namedParam, enumerable);
+						break;
+					default:
+						SetParameter(namedParam, obj, DetermineType(namedParam, obj.GetType()));
+						break;
+				}
+			}
+			return this;
+		}
+
+		// Since 5.3
+		[Obsolete("Use SetParameters method instead.")]
 		public IQuery SetProperties(object bean)
```

Comment:
What if we use this method to support the `IDictionary<string, object>`? 
```suggestion
		public IQuery SetProperties(object bean)
		{
			if (bean is IDictionary<string, object> map)
			{
			// TODO:
			}
		}
```

Reply:
It would be less discoverable, but it would support `dynamic` without having to do an explicit static call on some extension class.